File: Algorithms/Python/0017/0017.py
class Solution:
    res = []
    map = {"2": "abc", "3": "def", "4": "ghi", "5": "jkl", "6": "mno", "7": "pqrs", "8": "tuv", "9": "xyz"}
    def letterCombinations(self, digits: str) -> List[str]:
        def DFS(i: int, t: str):
            if i == len(digits):
                res.append(t)
            else:
                s = digits[i]
                for j in range(len(s)):
                    DFS(i + 1, t + s[j])

        if (len(digits) != 0):
            DFS(0, "")
        return res

File: Algorithms/C++/0017/0017.cpp
class Solution {
public:
    vector<string> res;
    vector<vector<char>> map = {
        {}, {}, {'a', 'b', 'c'}, {'d', 'e', 'f'},
        {'g', 'h', 'i'}, {'j', 'k', 'l'}, {'m', 'n', 'o'},
        {'p', 'q', 'r', 's'}, {'t', 'u', 'v'}, {'w', 'x', 'y', 'z'}
    };
    vector<string> letterCombinations(string digits) {
        if (!digits.empty()) {
            DFS(digits, 0, "");
        }
        return res;
    }
    void DFS(string digits, int i, string t) {
        if (i == digits.length()) {
            res.push_back(t);
            return;
        }
        for (char ch : map[digits[i] - '0']) {
            DFS(digits, i + 1, t + ch);
        }
    }
};
